<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Achaz von Hardenberg" />

<meta name="date" content="2025-12-07" />

<title>Alternative Covariance Structures</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Alternative Covariance Structures</h1>
<h4 class="author">Achaz von Hardenberg</h4>
<h4 class="date">2025-12-07</h4>



<div id="beyond-phylogeny-alternative-covariance-structures" class="section level1">
<h1>Beyond Phylogeny: Alternative Covariance Structures</h1>
<p>phybaseR’s <code>tree</code> argument accepts <strong>any covariance
structure</strong>, not just phylogenies. This enables modeling:</p>
<ul>
<li><strong>Spatial autocorrelation</strong> (geographic
dependencies)</li>
<li><strong>Pedigrees</strong> (animal models / quantitative
genetics)</li>
<li><strong>Social networks</strong> (behavioral dependencies)</li>
<li><strong>Temporal autocorrelation</strong> (time series)</li>
<li><strong>Custom similarity matrices</strong> (any correlation
structure)</li>
</ul>
<hr />
<div id="why-model-covariance-structure" class="section level2">
<h2>Why Model Covariance Structure?</h2>
<p>Similar to phylogenetic non-independence, other structures create
correlations among observations:</p>
<ul>
<li><strong>Spatial</strong>: Nearby sites are more similar</li>
<li><strong>Genetic</strong>: Related individuals share alleles</li>
<li><strong>Social</strong>: Connected individuals influence each
other</li>
<li><strong>Temporal</strong>: Adjacent time points are correlated</li>
</ul>
<p><strong>Ignoring structure</strong> → Inflated Type I error, wrong
estimates<br />
<strong>Modeling structure</strong> → Proper inference, correct
uncertainty</p>
<hr />
</div>
<div id="spatial-autocorrelation" class="section level2">
<h2>Spatial Autocorrelation</h2>
<p>Account for geographic dependencies in your data.</p>
<div id="basic-spatial-model" class="section level3">
<h3>Basic Spatial Model</h3>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(geosphere) <span class="co"># For distance calculations</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="co"># Data with coordinates</span></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a>    <span class="at">site_id =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">50</span>,</span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a>    <span class="at">latitude =</span> <span class="fu">runif</span>(<span class="dv">50</span>, <span class="dv">40</span>, <span class="dv">50</span>),</span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a>    <span class="at">longitude =</span> <span class="fu">runif</span>(<span class="dv">50</span>, <span class="sc">-</span><span class="dv">10</span>, <span class="dv">10</span>),</span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a>    <span class="at">temperature =</span> <span class="fu">rnorm</span>(<span class="dv">50</span>, <span class="dv">15</span>, <span class="dv">3</span>),</span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a>    <span class="at">species_richness =</span> <span class="fu">rpois</span>(<span class="dv">50</span>, <span class="dv">20</span>),</span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a>    <span class="at">habitat_quality =</span> <span class="fu">rnorm</span>(<span class="dv">50</span>, <span class="dv">0</span>, <span class="dv">1</span>)</span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a>)</span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a></span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a><span class="co"># Calculate geographic distances</span></span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a>coords <span class="ot">&lt;-</span> <span class="fu">cbind</span>(data<span class="sc">$</span>longitude, data<span class="sc">$</span>latitude)</span>
<span id="cb1-15"><a href="#cb1-15" tabindex="-1"></a>distances <span class="ot">&lt;-</span> <span class="fu">distm</span>(coords) <span class="sc">/</span> <span class="dv">1000</span> <span class="co"># Convert to km</span></span>
<span id="cb1-16"><a href="#cb1-16" tabindex="-1"></a></span>
<span id="cb1-17"><a href="#cb1-17" tabindex="-1"></a><span class="co"># Create spatial covariance matrix (exponential decay)</span></span>
<span id="cb1-18"><a href="#cb1-18" tabindex="-1"></a>spatial_sigma <span class="ot">&lt;-</span> <span class="dv">100</span> <span class="co"># Spatial scale (km)</span></span>
<span id="cb1-19"><a href="#cb1-19" tabindex="-1"></a>spatial_cov <span class="ot">&lt;-</span> <span class="fu">exp</span>(<span class="sc">-</span>distances <span class="sc">/</span> spatial_sigma)</span>
<span id="cb1-20"><a href="#cb1-20" tabindex="-1"></a></span>
<span id="cb1-21"><a href="#cb1-21" tabindex="-1"></a><span class="co"># Format for phybaseR</span></span>
<span id="cb1-22"><a href="#cb1-22" tabindex="-1"></a><span class="fu">rownames</span>(spatial_cov) <span class="ot">&lt;-</span> data<span class="sc">$</span>site_id</span>
<span id="cb1-23"><a href="#cb1-23" tabindex="-1"></a><span class="fu">colnames</span>(spatial_cov) <span class="ot">&lt;-</span> data<span class="sc">$</span>site_id</span>
<span id="cb1-24"><a href="#cb1-24" tabindex="-1"></a><span class="fu">rownames</span>(data) <span class="ot">&lt;-</span> data<span class="sc">$</span>site_id</span>
<span id="cb1-25"><a href="#cb1-25" tabindex="-1"></a></span>
<span id="cb1-26"><a href="#cb1-26" tabindex="-1"></a><span class="co"># Fit model with spatial structure</span></span>
<span id="cb1-27"><a href="#cb1-27" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">phybase_run</span>(</span>
<span id="cb1-28"><a href="#cb1-28" tabindex="-1"></a>    <span class="at">equations =</span> <span class="fu">list</span>(</span>
<span id="cb1-29"><a href="#cb1-29" tabindex="-1"></a>        species_richness <span class="sc">~</span> temperature <span class="sc">+</span> habitat_quality</span>
<span id="cb1-30"><a href="#cb1-30" tabindex="-1"></a>    ),</span>
<span id="cb1-31"><a href="#cb1-31" tabindex="-1"></a>    <span class="at">data =</span> data,</span>
<span id="cb1-32"><a href="#cb1-32" tabindex="-1"></a>    <span class="at">tree =</span> spatial_cov, <span class="co"># Spatial covariance!</span></span>
<span id="cb1-33"><a href="#cb1-33" tabindex="-1"></a>    <span class="at">dsep =</span> <span class="cn">TRUE</span>,</span>
<span id="cb1-34"><a href="#cb1-34" tabindex="-1"></a>    <span class="at">n.chains =</span> <span class="dv">3</span>,</span>
<span id="cb1-35"><a href="#cb1-35" tabindex="-1"></a>    <span class="at">n.iter =</span> <span class="dv">5000</span></span>
<span id="cb1-36"><a href="#cb1-36" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="spatial-correlation-functions" class="section level3">
<h3>Spatial Correlation Functions</h3>
<p>Different functions for different spatial patterns:</p>
<div id="exponential-decay" class="section level4">
<h4>Exponential Decay</h4>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="co"># Sharp initial decline, long tail</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a>spatial_cov <span class="ot">&lt;-</span> <span class="fu">exp</span>(<span class="sc">-</span>distances <span class="sc">/</span> sigma)</span></code></pre></div>
<p><strong>Use when</strong>: Gradual spatial gradients</p>
</div>
<div id="gaussian-decay" class="section level4">
<h4>Gaussian Decay</h4>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co"># Smooth, faster decay at distance</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>spatial_cov <span class="ot">&lt;-</span> <span class="fu">exp</span>(<span class="sc">-</span>(distances <span class="sc">/</span> sigma)<span class="sc">^</span><span class="dv">2</span>)</span></code></pre></div>
<p><strong>Use when</strong>: Strong local correlation, weak distant
correlation</p>
</div>
<div id="matérn-correlation" class="section level4">
<h4>Matérn Correlation</h4>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">library</span>(geoR)</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="co"># More flexible, controls smoothness</span></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a>spatial_cov <span class="ot">&lt;-</span> <span class="fu">matern</span>(distances, <span class="at">phi =</span> sigma, <span class="at">kappa =</span> <span class="fl">1.5</span>)</span></code></pre></div>
<p><strong>Use when</strong>: Need to control smoothness parameter</p>
</div>
</div>
<div id="choosing-spatial-scale-σ" class="section level3">
<h3>Choosing Spatial Scale (σ)</h3>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="co"># Exploratory: variogram</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="fu">library</span>(gstat)</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a><span class="fu">library</span>(sp)</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a></span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="fu">coordinates</span>(data) <span class="ot">&lt;-</span> <span class="er">~</span> longitude <span class="sc">+</span> latitude</span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a>variogram_emp <span class="ot">&lt;-</span> <span class="fu">variogram</span>(species_richness <span class="sc">~</span> <span class="dv">1</span>, data)</span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a><span class="fu">plot</span>(variogram_emp)</span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a></span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a><span class="co"># Fit variogram to estimate range</span></span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a>vgm_fit <span class="ot">&lt;-</span> <span class="fu">fit.variogram</span>(variogram_emp, <span class="fu">vgm</span>(<span class="st">&quot;Exp&quot;</span>))</span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a>sigma_estimate <span class="ot">&lt;-</span> vgm_fit<span class="sc">$</span>range[<span class="dv">2</span>]</span></code></pre></div>
<hr />
</div>
</div>
<div id="pedigrees-animal-models" class="section level2">
<h2>Pedigrees (Animal Models)</h2>
<p>Quantitative genetics using pedigree-based relatedness.</p>
<div id="basic-animal-model" class="section level3">
<h3>Basic Animal Model</h3>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">library</span>(kinship2) <span class="co"># Or nadiv, pedigreemm</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a></span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="co"># Pedigree structure</span></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a>pedigree <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a>    <span class="at">id =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">100</span>,</span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a>    <span class="at">sire =</span> <span class="fu">c</span>(<span class="cn">NA</span>, <span class="cn">NA</span>, <span class="fu">rep</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="at">each =</span> <span class="dv">10</span>), ...), <span class="co"># Father IDs</span></span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a>    <span class="at">dam =</span> <span class="fu">c</span>(<span class="cn">NA</span>, <span class="cn">NA</span>, <span class="fu">rep</span>(<span class="dv">11</span><span class="sc">:</span><span class="dv">20</span>, <span class="at">each =</span> <span class="dv">10</span>), ...) <span class="co"># Mother IDs</span></span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a>)</span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a></span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a><span class="co"># Phenotype data</span></span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a>    <span class="at">individual =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">100</span>,</span>
<span id="cb6-13"><a href="#cb6-13" tabindex="-1"></a>    <span class="at">body_size =</span> <span class="fu">rnorm</span>(<span class="dv">100</span>, <span class="dv">50</span>, <span class="dv">10</span>),</span>
<span id="cb6-14"><a href="#cb6-14" tabindex="-1"></a>    <span class="at">litter_size =</span> <span class="fu">rpois</span>(<span class="dv">100</span>, <span class="dv">5</span>),</span>
<span id="cb6-15"><a href="#cb6-15" tabindex="-1"></a>    <span class="at">maternal_care =</span> <span class="fu">rnorm</span>(<span class="dv">100</span>, <span class="dv">0</span>, <span class="dv">1</span>),</span>
<span id="cb6-16"><a href="#cb6-16" tabindex="-1"></a>    <span class="at">food_availability =</span> <span class="fu">rnorm</span>(<span class="dv">100</span>, <span class="dv">0</span>, <span class="dv">1</span>)</span>
<span id="cb6-17"><a href="#cb6-17" tabindex="-1"></a>)</span>
<span id="cb6-18"><a href="#cb6-18" tabindex="-1"></a></span>
<span id="cb6-19"><a href="#cb6-19" tabindex="-1"></a><span class="co"># Calculate additive genetic relatedness matrix (A-matrix)</span></span>
<span id="cb6-20"><a href="#cb6-20" tabindex="-1"></a>ped_obj <span class="ot">&lt;-</span> <span class="fu">pedigree</span>(</span>
<span id="cb6-21"><a href="#cb6-21" tabindex="-1"></a>    <span class="at">id =</span> pedigree<span class="sc">$</span>id,</span>
<span id="cb6-22"><a href="#cb6-22" tabindex="-1"></a>    <span class="at">dadid =</span> pedigree<span class="sc">$</span>sire,</span>
<span id="cb6-23"><a href="#cb6-23" tabindex="-1"></a>    <span class="at">momid =</span> pedigree<span class="sc">$</span>dam</span>
<span id="cb6-24"><a href="#cb6-24" tabindex="-1"></a>)</span>
<span id="cb6-25"><a href="#cb6-25" tabindex="-1"></a>A_matrix <span class="ot">&lt;-</span> <span class="fu">kinship</span>(ped_obj) <span class="sc">*</span> <span class="dv">2</span> <span class="co"># A = 2 × kinship</span></span>
<span id="cb6-26"><a href="#cb6-26" tabindex="-1"></a></span>
<span id="cb6-27"><a href="#cb6-27" tabindex="-1"></a><span class="co"># Ensure matching</span></span>
<span id="cb6-28"><a href="#cb6-28" tabindex="-1"></a><span class="fu">rownames</span>(A_matrix) <span class="ot">&lt;-</span> pedigree<span class="sc">$</span>id</span>
<span id="cb6-29"><a href="#cb6-29" tabindex="-1"></a><span class="fu">colnames</span>(A_matrix) <span class="ot">&lt;-</span> pedigree<span class="sc">$</span>id</span>
<span id="cb6-30"><a href="#cb6-30" tabindex="-1"></a><span class="fu">rownames</span>(data) <span class="ot">&lt;-</span> data<span class="sc">$</span>individual</span>
<span id="cb6-31"><a href="#cb6-31" tabindex="-1"></a></span>
<span id="cb6-32"><a href="#cb6-32" tabindex="-1"></a><span class="co"># Quantitative genetics SEM!</span></span>
<span id="cb6-33"><a href="#cb6-33" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">phybase_run</span>(</span>
<span id="cb6-34"><a href="#cb6-34" tabindex="-1"></a>    <span class="at">equations =</span> <span class="fu">list</span>(</span>
<span id="cb6-35"><a href="#cb6-35" tabindex="-1"></a>        body_size <span class="sc">~</span> food_availability <span class="sc">+</span> maternal_care,</span>
<span id="cb6-36"><a href="#cb6-36" tabindex="-1"></a>        litter_size <span class="sc">~</span> body_size <span class="sc">+</span> maternal_care</span>
<span id="cb6-37"><a href="#cb6-37" tabindex="-1"></a>    ),</span>
<span id="cb6-38"><a href="#cb6-38" tabindex="-1"></a>    <span class="at">data =</span> data,</span>
<span id="cb6-39"><a href="#cb6-39" tabindex="-1"></a>    <span class="at">tree =</span> A_matrix, <span class="co"># Genetic relatedness</span></span>
<span id="cb6-40"><a href="#cb6-40" tabindex="-1"></a>    <span class="at">dsep =</span> <span class="cn">TRUE</span>,</span>
<span id="cb6-41"><a href="#cb6-41" tabindex="-1"></a>    <span class="at">n.chains =</span> <span class="dv">3</span>,</span>
<span id="cb6-42"><a href="#cb6-42" tabindex="-1"></a>    <span class="at">n.iter =</span> <span class="dv">5000</span></span>
<span id="cb6-43"><a href="#cb6-43" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="what-the-a-matrix-captures" class="section level3">
<h3>What the A-Matrix Captures</h3>
<ul>
<li><strong>Diagonal</strong>: 1 (individual’s relatedness to
itself)</li>
<li><strong>Parent-offspring</strong>: 0.5 (share 50% genes)</li>
<li><strong>Full siblings</strong>: 0.5 on average</li>
<li><strong>Half siblings</strong>: 0.25</li>
<li><strong>Unrelated</strong>: 0</li>
</ul>
</div>
<div id="applications" class="section level3">
<h3>Applications</h3>
<p><strong>Heritability estimation</strong>:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="co"># Variance components from model</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a><span class="co"># h² = V_genetic / (V_genetic + V_residual)</span></span></code></pre></div>
<p><strong>Breeding values</strong> for selection:</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a><span class="co"># Genetic merit of individuals</span></span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a><span class="co"># Used in animal/plant breeding</span></span></code></pre></div>
<p><strong>Genetic correlations</strong>:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="co"># Are traits genetically correlated?</span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">phybase_run</span>(</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a>    <span class="at">equations =</span> <span class="fu">list</span>(</span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a>        trait1 <span class="sc">~</span> environment,</span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a>        trait2 <span class="sc">~</span> environment</span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a>    ),</span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a>    <span class="at">data =</span> data,</span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a>    <span class="at">tree =</span> A_matrix</span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a>)</span>
<span id="cb9-10"><a href="#cb9-10" tabindex="-1"></a><span class="co"># Correlation in genetic effects</span></span></code></pre></div>
</div>
<div id="advanced-dominance-epistasis" class="section level3">
<h3>Advanced: Dominance &amp; Epistasis</h3>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="co"># Additive + dominance</span></span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a><span class="fu">library</span>(nadiv)</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>D_matrix <span class="ot">&lt;-</span> <span class="fu">makeD</span>(pedigree) <span class="co"># Dominance relatedness</span></span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a></span>
<span id="cb10-5"><a href="#cb10-5" tabindex="-1"></a><span class="co"># Would need custom JAGS to include both A and D</span></span></code></pre></div>
<hr />
</div>
</div>
<div id="social-networks" class="section level2">
<h2>Social Networks</h2>
<p>Model dependencies from social structure.</p>
<div id="network-based-covariance" class="section level3">
<h3>Network-Based Covariance</h3>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="fu">library</span>(igraph)</span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a></span>
<span id="cb11-3"><a href="#cb11-3" tabindex="-1"></a><span class="co"># Social network adjacency matrix</span></span>
<span id="cb11-4"><a href="#cb11-4" tabindex="-1"></a><span class="co"># 1 = individuals interact, 0 = no interaction</span></span>
<span id="cb11-5"><a href="#cb11-5" tabindex="-1"></a>social_adj <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="dv">0</span>, <span class="dv">30</span>, <span class="dv">30</span>)</span>
<span id="cb11-6"><a href="#cb11-6" tabindex="-1"></a><span class="co"># Fill with observed interactions</span></span>
<span id="cb11-7"><a href="#cb11-7" tabindex="-1"></a>social_adj[<span class="dv">1</span>, <span class="dv">2</span>] <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb11-8"><a href="#cb11-8" tabindex="-1"></a>social_adj[<span class="dv">2</span>, <span class="dv">1</span>] <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb11-9"><a href="#cb11-9" tabindex="-1"></a><span class="co"># ... etc</span></span>
<span id="cb11-10"><a href="#cb11-10" tabindex="-1"></a></span>
<span id="cb11-11"><a href="#cb11-11" tabindex="-1"></a><span class="co"># Convert to covariance</span></span>
<span id="cb11-12"><a href="#cb11-12" tabindex="-1"></a><span class="co"># Option 1: Direct adjacency as correlation</span></span>
<span id="cb11-13"><a href="#cb11-13" tabindex="-1"></a>social_cov <span class="ot">&lt;-</span> social_adj</span>
<span id="cb11-14"><a href="#cb11-14" tabindex="-1"></a><span class="fu">diag</span>(social_cov) <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb11-15"><a href="#cb11-15" tabindex="-1"></a></span>
<span id="cb11-16"><a href="#cb11-16" tabindex="-1"></a><span class="co"># Option 2: Normalized Laplacian</span></span>
<span id="cb11-17"><a href="#cb11-17" tabindex="-1"></a>graph <span class="ot">&lt;-</span> <span class="fu">graph_from_adjacency_matrix</span>(social_adj, <span class="at">mode =</span> <span class="st">&quot;undirected&quot;</span>)</span>
<span id="cb11-18"><a href="#cb11-18" tabindex="-1"></a>laplacian <span class="ot">&lt;-</span> <span class="fu">laplacian_matrix</span>(graph, <span class="at">normalized =</span> <span class="cn">TRUE</span>)</span>
<span id="cb11-19"><a href="#cb11-19" tabindex="-1"></a>social_cov <span class="ot">&lt;-</span> <span class="fu">solve</span>(laplacian <span class="sc">+</span> <span class="fu">diag</span>(<span class="fl">0.1</span>, <span class="fu">nrow</span>(laplacian))) <span class="co"># Regularize</span></span>
<span id="cb11-20"><a href="#cb11-20" tabindex="-1"></a></span>
<span id="cb11-21"><a href="#cb11-21" tabindex="-1"></a><span class="co"># Behavioral data</span></span>
<span id="cb11-22"><a href="#cb11-22" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb11-23"><a href="#cb11-23" tabindex="-1"></a>    <span class="at">individual =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>,</span>
<span id="cb11-24"><a href="#cb11-24" tabindex="-1"></a>    <span class="at">dominance_rank =</span> <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">30</span>),</span>
<span id="cb11-25"><a href="#cb11-25" tabindex="-1"></a>    <span class="at">mating_success =</span> <span class="fu">rpois</span>(<span class="dv">30</span>, <span class="dv">3</span>),</span>
<span id="cb11-26"><a href="#cb11-26" tabindex="-1"></a>    <span class="at">foraging_efficiency =</span> <span class="fu">rnorm</span>(<span class="dv">30</span>, <span class="dv">100</span>, <span class="dv">20</span>),</span>
<span id="cb11-27"><a href="#cb11-27" tabindex="-1"></a>    <span class="at">aggression =</span> <span class="fu">rnorm</span>(<span class="dv">30</span>, <span class="dv">0</span>, <span class="dv">1</span>)</span>
<span id="cb11-28"><a href="#cb11-28" tabindex="-1"></a>)</span>
<span id="cb11-29"><a href="#cb11-29" tabindex="-1"></a></span>
<span id="cb11-30"><a href="#cb11-30" tabindex="-1"></a><span class="fu">rownames</span>(social_cov) <span class="ot">&lt;-</span> data<span class="sc">$</span>individual</span>
<span id="cb11-31"><a href="#cb11-31" tabindex="-1"></a><span class="fu">colnames</span>(social_cov) <span class="ot">&lt;-</span> data<span class="sc">$</span>individual</span>
<span id="cb11-32"><a href="#cb11-32" tabindex="-1"></a><span class="fu">rownames</span>(data) <span class="ot">&lt;-</span> data<span class="sc">$</span>individual</span>
<span id="cb11-33"><a href="#cb11-33" tabindex="-1"></a></span>
<span id="cb11-34"><a href="#cb11-34" tabindex="-1"></a><span class="co"># SEM with social structure</span></span>
<span id="cb11-35"><a href="#cb11-35" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">phybase_run</span>(</span>
<span id="cb11-36"><a href="#cb11-36" tabindex="-1"></a>    <span class="at">equations =</span> <span class="fu">list</span>(</span>
<span id="cb11-37"><a href="#cb11-37" tabindex="-1"></a>        mating_success <span class="sc">~</span> dominance_rank <span class="sc">+</span> aggression,</span>
<span id="cb11-38"><a href="#cb11-38" tabindex="-1"></a>        foraging_efficiency <span class="sc">~</span> dominance_rank</span>
<span id="cb11-39"><a href="#cb11-39" tabindex="-1"></a>    ),</span>
<span id="cb11-40"><a href="#cb11-40" tabindex="-1"></a>    <span class="at">data =</span> data,</span>
<span id="cb11-41"><a href="#cb11-41" tabindex="-1"></a>    <span class="at">tree =</span> social_cov,</span>
<span id="cb11-42"><a href="#cb11-42" tabindex="-1"></a>    <span class="at">n.chains =</span> <span class="dv">3</span>,</span>
<span id="cb11-43"><a href="#cb11-43" tabindex="-1"></a>    <span class="at">n.iter =</span> <span class="dv">5000</span></span>
<span id="cb11-44"><a href="#cb11-44" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="network-metrics-as-covariates" class="section level3">
<h3>Network Metrics as Covariates</h3>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="co"># Instead of (or in addition to) covariance structure</span></span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a>g <span class="ot">&lt;-</span> <span class="fu">graph_from_adjacency_matrix</span>(social_adj)</span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a></span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a>data<span class="sc">$</span>degree <span class="ot">&lt;-</span> <span class="fu">degree</span>(g) <span class="co"># Number of connections</span></span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a>data<span class="sc">$</span>betweenness <span class="ot">&lt;-</span> <span class="fu">betweenness</span>(g) <span class="co"># Bridging position</span></span>
<span id="cb12-6"><a href="#cb12-6" tabindex="-1"></a>data<span class="sc">$</span>eigenvector <span class="ot">&lt;-</span> <span class="fu">eigen_centrality</span>(g)<span class="sc">$</span>vector <span class="co"># Influence</span></span>
<span id="cb12-7"><a href="#cb12-7" tabindex="-1"></a></span>
<span id="cb12-8"><a href="#cb12-8" tabindex="-1"></a><span class="co"># Use as predictors</span></span>
<span id="cb12-9"><a href="#cb12-9" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">phybase_run</span>(</span>
<span id="cb12-10"><a href="#cb12-10" tabindex="-1"></a>    <span class="at">equations =</span> <span class="fu">list</span>(</span>
<span id="cb12-11"><a href="#cb12-11" tabindex="-1"></a>        mating_success <span class="sc">~</span> degree <span class="sc">+</span> betweenness</span>
<span id="cb12-12"><a href="#cb12-12" tabindex="-1"></a>    ),</span>
<span id="cb12-13"><a href="#cb12-13" tabindex="-1"></a>    <span class="at">data =</span> data,</span>
<span id="cb12-14"><a href="#cb12-14" tabindex="-1"></a>    <span class="at">tree =</span> social_cov <span class="co"># Still account for network structure</span></span>
<span id="cb12-15"><a href="#cb12-15" tabindex="-1"></a>)</span></code></pre></div>
<hr />
</div>
</div>
<div id="temporal-autocorrelation" class="section level2">
<h2>Temporal Autocorrelation</h2>
<p>For time series or repeated measures over time.</p>
<div id="ar1-correlation-structure" class="section level3">
<h3>AR(1) Correlation Structure</h3>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="co"># Data ordered by time</span></span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb13-3"><a href="#cb13-3" tabindex="-1"></a>    <span class="at">time =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">100</span>,</span>
<span id="cb13-4"><a href="#cb13-4" tabindex="-1"></a>    <span class="at">measurement =</span> <span class="fu">arima.sim</span>(<span class="fu">list</span>(<span class="at">ar =</span> <span class="fl">0.7</span>), <span class="at">n =</span> <span class="dv">100</span>),</span>
<span id="cb13-5"><a href="#cb13-5" tabindex="-1"></a>    <span class="at">temperature =</span> <span class="fu">rnorm</span>(<span class="dv">100</span>, <span class="dv">15</span>, <span class="dv">3</span>),</span>
<span id="cb13-6"><a href="#cb13-6" tabindex="-1"></a>    <span class="at">rainfall =</span> <span class="fu">rnorm</span>(<span class="dv">100</span>, <span class="dv">50</span>, <span class="dv">20</span>)</span>
<span id="cb13-7"><a href="#cb13-7" tabindex="-1"></a>)</span>
<span id="cb13-8"><a href="#cb13-8" tabindex="-1"></a></span>
<span id="cb13-9"><a href="#cb13-9" tabindex="-1"></a><span class="co"># AR(1) correlation matrix</span></span>
<span id="cb13-10"><a href="#cb13-10" tabindex="-1"></a>rho <span class="ot">&lt;-</span> <span class="fl">0.7</span> <span class="co"># Autocorrelation parameter</span></span>
<span id="cb13-11"><a href="#cb13-11" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="fu">nrow</span>(data)</span>
<span id="cb13-12"><a href="#cb13-12" tabindex="-1"></a>temporal_cov <span class="ot">&lt;-</span> rho<span class="sc">^</span><span class="fu">abs</span>(<span class="fu">outer</span>(<span class="dv">1</span><span class="sc">:</span>n, <span class="dv">1</span><span class="sc">:</span>n, <span class="st">&quot;-&quot;</span>))</span>
<span id="cb13-13"><a href="#cb13-13" tabindex="-1"></a></span>
<span id="cb13-14"><a href="#cb13-14" tabindex="-1"></a><span class="fu">rownames</span>(temporal_cov) <span class="ot">&lt;-</span> data<span class="sc">$</span>time</span>
<span id="cb13-15"><a href="#cb13-15" tabindex="-1"></a><span class="fu">colnames</span>(temporal_cov) <span class="ot">&lt;-</span> data<span class="sc">$</span>time</span>
<span id="cb13-16"><a href="#cb13-16" tabindex="-1"></a><span class="fu">rownames</span>(data) <span class="ot">&lt;-</span> data<span class="sc">$</span>time</span>
<span id="cb13-17"><a href="#cb13-17" tabindex="-1"></a></span>
<span id="cb13-18"><a href="#cb13-18" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">phybase_run</span>(</span>
<span id="cb13-19"><a href="#cb13-19" tabindex="-1"></a>    <span class="at">equations =</span> <span class="fu">list</span>(</span>
<span id="cb13-20"><a href="#cb13-20" tabindex="-1"></a>        measurement <span class="sc">~</span> temperature <span class="sc">+</span> rainfall</span>
<span id="cb13-21"><a href="#cb13-21" tabindex="-1"></a>    ),</span>
<span id="cb13-22"><a href="#cb13-22" tabindex="-1"></a>    <span class="at">data =</span> data,</span>
<span id="cb13-23"><a href="#cb13-23" tabindex="-1"></a>    <span class="at">tree =</span> temporal_cov,</span>
<span id="cb13-24"><a href="#cb13-24" tabindex="-1"></a>    <span class="at">n.chains =</span> <span class="dv">3</span>,</span>
<span id="cb13-25"><a href="#cb13-25" tabindex="-1"></a>    <span class="at">n.iter =</span> <span class="dv">5000</span></span>
<span id="cb13-26"><a href="#cb13-26" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="estimating-ρ" class="section level3">
<h3>Estimating ρ</h3>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="co"># From data</span></span>
<span id="cb14-2"><a href="#cb14-2" tabindex="-1"></a><span class="fu">acf</span>(data<span class="sc">$</span>measurement, <span class="at">plot =</span> <span class="cn">FALSE</span>)<span class="sc">$</span>acf[<span class="dv">2</span>] <span class="co"># Lag-1 autocorrelation</span></span>
<span id="cb14-3"><a href="#cb14-3" tabindex="-1"></a></span>
<span id="cb14-4"><a href="#cb14-4" tabindex="-1"></a><span class="co"># Or use nlme/lme4 first</span></span>
<span id="cb14-5"><a href="#cb14-5" tabindex="-1"></a><span class="fu">library</span>(nlme)</span>
<span id="cb14-6"><a href="#cb14-6" tabindex="-1"></a>temp_fit <span class="ot">&lt;-</span> <span class="fu">gls</span>(measurement <span class="sc">~</span> temperature <span class="sc">+</span> rainfall,</span>
<span id="cb14-7"><a href="#cb14-7" tabindex="-1"></a>    <span class="at">correlation =</span> <span class="fu">corAR1</span>(<span class="at">form =</span> <span class="sc">~</span>time),</span>
<span id="cb14-8"><a href="#cb14-8" tabindex="-1"></a>    <span class="at">data =</span> data</span>
<span id="cb14-9"><a href="#cb14-9" tabindex="-1"></a>)</span>
<span id="cb14-10"><a href="#cb14-10" tabindex="-1"></a>rho_est <span class="ot">&lt;-</span> <span class="fu">coef</span>(temp_fit<span class="sc">$</span>modelStruct<span class="sc">$</span>corStruct, <span class="at">unconstrained =</span> <span class="cn">FALSE</span>)</span></code></pre></div>
<hr />
</div>
</div>
<div id="custom-similarity-matrices" class="section level2">
<h2>Custom Similarity Matrices</h2>
<p>Any symmetric, positive semi-definite matrix works.</p>
<div id="functional-trait-similarity" class="section level3">
<h3>Functional Trait Similarity</h3>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="co"># Species functional traits</span></span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a>traits <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb15-3"><a href="#cb15-3" tabindex="-1"></a>    <span class="at">species =</span> <span class="fu">paste0</span>(<span class="st">&quot;sp&quot;</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">50</span>),</span>
<span id="cb15-4"><a href="#cb15-4" tabindex="-1"></a>    <span class="at">bill_length =</span> <span class="fu">rnorm</span>(<span class="dv">50</span>, <span class="dv">20</span>, <span class="dv">5</span>),</span>
<span id="cb15-5"><a href="#cb15-5" tabindex="-1"></a>    <span class="at">wing_length =</span> <span class="fu">rnorm</span>(<span class="dv">50</span>, <span class="dv">100</span>, <span class="dv">20</span>),</span>
<span id="cb15-6"><a href="#cb15-6" tabindex="-1"></a>    <span class="at">body_mass =</span> <span class="fu">rnorm</span>(<span class="dv">50</span>, <span class="dv">50</span>, <span class="dv">15</span>)</span>
<span id="cb15-7"><a href="#cb15-7" tabindex="-1"></a>)</span>
<span id="cb15-8"><a href="#cb15-8" tabindex="-1"></a></span>
<span id="cb15-9"><a href="#cb15-9" tabindex="-1"></a><span class="co"># Euclidean distance in trait space</span></span>
<span id="cb15-10"><a href="#cb15-10" tabindex="-1"></a>trait_matrix <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(traits[, <span class="sc">-</span><span class="dv">1</span>])</span>
<span id="cb15-11"><a href="#cb15-11" tabindex="-1"></a>trait_dist <span class="ot">&lt;-</span> <span class="fu">dist</span>(trait_matrix)</span>
<span id="cb15-12"><a href="#cb15-12" tabindex="-1"></a>trait_dist_matrix <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(trait_dist)</span>
<span id="cb15-13"><a href="#cb15-13" tabindex="-1"></a></span>
<span id="cb15-14"><a href="#cb15-14" tabindex="-1"></a><span class="co"># Convert distance to similarity (Gaussian kernel)</span></span>
<span id="cb15-15"><a href="#cb15-15" tabindex="-1"></a>sigma_trait <span class="ot">&lt;-</span> <span class="fu">median</span>(trait_dist_matrix)</span>
<span id="cb15-16"><a href="#cb15-16" tabindex="-1"></a>functional_sim <span class="ot">&lt;-</span> <span class="fu">exp</span>(<span class="sc">-</span>(trait_dist_matrix <span class="sc">/</span> sigma_trait)<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb15-17"><a href="#cb15-17" tabindex="-1"></a></span>
<span id="cb15-18"><a href="#cb15-18" tabindex="-1"></a><span class="fu">rownames</span>(functional_sim) <span class="ot">&lt;-</span> traits<span class="sc">$</span>species</span>
<span id="cb15-19"><a href="#cb15-19" tabindex="-1"></a><span class="fu">colnames</span>(functional_sim) <span class="ot">&lt;-</span> traits<span class="sc">$</span>species</span>
<span id="cb15-20"><a href="#cb15-20" tabindex="-1"></a></span>
<span id="cb15-21"><a href="#cb15-21" tabindex="-1"></a><span class="co"># Use in SEM</span></span>
<span id="cb15-22"><a href="#cb15-22" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">phybase_run</span>(</span>
<span id="cb15-23"><a href="#cb15-23" tabindex="-1"></a>    <span class="at">equations =</span> equations,</span>
<span id="cb15-24"><a href="#cb15-24" tabindex="-1"></a>    <span class="at">data =</span> data,</span>
<span id="cb15-25"><a href="#cb15-25" tabindex="-1"></a>    <span class="at">tree =</span> functional_sim,</span>
<span id="cb15-26"><a href="#cb15-26" tabindex="-1"></a>    <span class="at">n.chains =</span> <span class="dv">3</span>,</span>
<span id="cb15-27"><a href="#cb15-27" tabindex="-1"></a>    <span class="at">n.iter =</span> <span class="dv">5000</span></span>
<span id="cb15-28"><a href="#cb15-28" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="genomic-similarity-snps" class="section level3">
<h3>Genomic Similarity (SNPs)</h3>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a><span class="co"># Genomic relationship matrix from SNPs</span></span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a><span class="fu">library</span>(rrBLUP)</span>
<span id="cb16-3"><a href="#cb16-3" tabindex="-1"></a></span>
<span id="cb16-4"><a href="#cb16-4" tabindex="-1"></a><span class="co"># SNP data (individuals × markers)</span></span>
<span id="cb16-5"><a href="#cb16-5" tabindex="-1"></a>snps <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">sample</span>(<span class="dv">0</span><span class="sc">:</span><span class="dv">2</span>, <span class="dv">100</span> <span class="sc">*</span> <span class="dv">1000</span>, <span class="at">replace =</span> <span class="cn">TRUE</span>), <span class="dv">100</span>, <span class="dv">1000</span>)</span>
<span id="cb16-6"><a href="#cb16-6" tabindex="-1"></a><span class="fu">rownames</span>(snps) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(<span class="st">&quot;ind&quot;</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">100</span>)</span>
<span id="cb16-7"><a href="#cb16-7" tabindex="-1"></a></span>
<span id="cb16-8"><a href="#cb16-8" tabindex="-1"></a><span class="co"># Calculate genomic relationship matrix</span></span>
<span id="cb16-9"><a href="#cb16-9" tabindex="-1"></a>G_matrix <span class="ot">&lt;-</span> <span class="fu">A.mat</span>(snps <span class="sc">-</span> <span class="dv">1</span>) <span class="co"># Center &amp; scale</span></span>
<span id="cb16-10"><a href="#cb16-10" tabindex="-1"></a></span>
<span id="cb16-11"><a href="#cb16-11" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">phybase_run</span>(</span>
<span id="cb16-12"><a href="#cb16-12" tabindex="-1"></a>    <span class="at">equations =</span> equations,</span>
<span id="cb16-13"><a href="#cb16-13" tabindex="-1"></a>    <span class="at">data =</span> data,</span>
<span id="cb16-14"><a href="#cb16-14" tabindex="-1"></a>    <span class="at">tree =</span> G_matrix,</span>
<span id="cb16-15"><a href="#cb16-15" tabindex="-1"></a>    <span class="at">n.chains =</span> <span class="dv">3</span>,</span>
<span id="cb16-16"><a href="#cb16-16" tabindex="-1"></a>    <span class="at">n.iter =</span> <span class="dv">5000</span></span>
<span id="cb16-17"><a href="#cb16-17" tabindex="-1"></a>)</span></code></pre></div>
<hr />
</div>
</div>
<div id="combining-multiple-structures" class="section level2">
<h2>Combining Multiple Structures</h2>
<div id="phylogeny-spatial" class="section level3">
<h3>Phylogeny + Spatial</h3>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="co"># Can&#39;t pass two matrices to tree argument</span></span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a><span class="co"># Use one as tree, other as random effect</span></span>
<span id="cb17-3"><a href="#cb17-3" tabindex="-1"></a></span>
<span id="cb17-4"><a href="#cb17-4" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">phybase_run</span>(</span>
<span id="cb17-5"><a href="#cb17-5" tabindex="-1"></a>    <span class="at">equations =</span> <span class="fu">list</span>(</span>
<span id="cb17-6"><a href="#cb17-6" tabindex="-1"></a>        trait <span class="sc">~</span> environment <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">|</span> site)</span>
<span id="cb17-7"><a href="#cb17-7" tabindex="-1"></a>    ),</span>
<span id="cb17-8"><a href="#cb17-8" tabindex="-1"></a>    <span class="at">data =</span> data,</span>
<span id="cb17-9"><a href="#cb17-9" tabindex="-1"></a>    <span class="at">tree =</span> phylo_tree, <span class="co"># Phylogenetic structure</span></span>
<span id="cb17-10"><a href="#cb17-10" tabindex="-1"></a>    <span class="at">random =</span> <span class="sc">~</span> (<span class="dv">1</span> <span class="sc">|</span> site), <span class="co"># Spatial grouping</span></span>
<span id="cb17-11"><a href="#cb17-11" tabindex="-1"></a>    <span class="at">n.chains =</span> <span class="dv">3</span>,</span>
<span id="cb17-12"><a href="#cb17-12" tabindex="-1"></a>    <span class="at">n.iter =</span> <span class="dv">5000</span></span>
<span id="cb17-13"><a href="#cb17-13" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="multiple-random-effects" class="section level3">
<h3>Multiple Random Effects</h3>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="co"># Phylogeny + Genetic + Spatial</span></span>
<span id="cb18-2"><a href="#cb18-2" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">phybase_run</span>(</span>
<span id="cb18-3"><a href="#cb18-3" tabindex="-1"></a>    <span class="at">equations =</span> <span class="fu">list</span>(</span>
<span id="cb18-4"><a href="#cb18-4" tabindex="-1"></a>        trait <span class="sc">~</span> environment <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">|</span> population) <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">|</span> site)</span>
<span id="cb18-5"><a href="#cb18-5" tabindex="-1"></a>    ),</span>
<span id="cb18-6"><a href="#cb18-6" tabindex="-1"></a>    <span class="at">data =</span> data,</span>
<span id="cb18-7"><a href="#cb18-7" tabindex="-1"></a>    <span class="at">tree =</span> phylo_tree, <span class="co"># Species phylogeny</span></span>
<span id="cb18-8"><a href="#cb18-8" tabindex="-1"></a>    <span class="at">random =</span> <span class="sc">~</span> (<span class="dv">1</span> <span class="sc">|</span> population) <span class="sc">+</span> (<span class="dv">1</span> <span class="sc">|</span> site), <span class="co"># Other structures</span></span>
<span id="cb18-9"><a href="#cb18-9" tabindex="-1"></a>    <span class="at">n.chains =</span> <span class="dv">3</span>,</span>
<span id="cb18-10"><a href="#cb18-10" tabindex="-1"></a>    <span class="at">n.iter =</span> <span class="dv">5000</span></span>
<span id="cb18-11"><a href="#cb18-11" tabindex="-1"></a>)</span></code></pre></div>
<hr />
</div>
</div>
<div id="requirements-for-custom-matrices" class="section level2">
<h2>Requirements for Custom Matrices</h2>
<p>Your covariance matrix must satisfy:</p>
<p>✅ <strong>Symmetric</strong>:
<code>all(cov_matrix == t(cov_matrix))</code><br />
✅ <strong>Positive semi-definite</strong>: All eigenvalues ≥ 0<br />
✅ <strong>Proper dimensions</strong>: n × n for n observations<br />
✅ <strong>Named rows/columns</strong>: Match data identifiers<br />
✅ <strong>Appropriate scale</strong>: Typically correlation matrix
(diagonal = 1)</p>
<div id="validation-checks" class="section level3">
<h3>Validation Checks</h3>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a>validate_covariance_matrix <span class="ot">&lt;-</span> <span class="cf">function</span>(mat, data) {</span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a>    <span class="co"># Symmetry</span></span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a>    <span class="cf">if</span> (<span class="sc">!</span><span class="fu">isSymmetric</span>(mat)) {</span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a>        <span class="fu">stop</span>(<span class="st">&quot;Matrix is not symmetric&quot;</span>)</span>
<span id="cb19-5"><a href="#cb19-5" tabindex="-1"></a>    }</span>
<span id="cb19-6"><a href="#cb19-6" tabindex="-1"></a></span>
<span id="cb19-7"><a href="#cb19-7" tabindex="-1"></a>    <span class="co"># Positive semi-definite</span></span>
<span id="cb19-8"><a href="#cb19-8" tabindex="-1"></a>    eigs <span class="ot">&lt;-</span> <span class="fu">eigen</span>(mat, <span class="at">only.values =</span> <span class="cn">TRUE</span>)<span class="sc">$</span>values</span>
<span id="cb19-9"><a href="#cb19-9" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">any</span>(eigs <span class="sc">&lt;</span> <span class="sc">-</span><span class="fl">1e-8</span>)) { <span class="co"># Allow tiny numerical error</span></span>
<span id="cb19-10"><a href="#cb19-10" tabindex="-1"></a>        <span class="fu">warning</span>(<span class="st">&quot;Matrix may not be positive semi-definite&quot;</span>)</span>
<span id="cb19-11"><a href="#cb19-11" tabindex="-1"></a>    }</span>
<span id="cb19-12"><a href="#cb19-12" tabindex="-1"></a></span>
<span id="cb19-13"><a href="#cb19-13" tabindex="-1"></a>    <span class="co"># Dimensions</span></span>
<span id="cb19-14"><a href="#cb19-14" tabindex="-1"></a>    <span class="cf">if</span> (<span class="fu">nrow</span>(mat) <span class="sc">!=</span> <span class="fu">nrow</span>(data)) {</span>
<span id="cb19-15"><a href="#cb19-15" tabindex="-1"></a>        <span class="fu">stop</span>(<span class="st">&quot;Matrix dimensions don&#39;t match data&quot;</span>)</span>
<span id="cb19-16"><a href="#cb19-16" tabindex="-1"></a>    }</span>
<span id="cb19-17"><a href="#cb19-17" tabindex="-1"></a></span>
<span id="cb19-18"><a href="#cb19-18" tabindex="-1"></a>    <span class="co"># Names</span></span>
<span id="cb19-19"><a href="#cb19-19" tabindex="-1"></a>    <span class="cf">if</span> (<span class="sc">!</span><span class="fu">all</span>(<span class="fu">rownames</span>(mat) <span class="sc">%in%</span> <span class="fu">rownames</span>(data))) {</span>
<span id="cb19-20"><a href="#cb19-20" tabindex="-1"></a>        <span class="fu">stop</span>(<span class="st">&quot;Matrix rownames don&#39;t match data rownames&quot;</span>)</span>
<span id="cb19-21"><a href="#cb19-21" tabindex="-1"></a>    }</span>
<span id="cb19-22"><a href="#cb19-22" tabindex="-1"></a></span>
<span id="cb19-23"><a href="#cb19-23" tabindex="-1"></a>    <span class="fu">message</span>(<span class="st">&quot;✓ Matrix passed validation&quot;</span>)</span>
<span id="cb19-24"><a href="#cb19-24" tabindex="-1"></a>    <span class="fu">return</span>(<span class="cn">TRUE</span>)</span>
<span id="cb19-25"><a href="#cb19-25" tabindex="-1"></a>}</span>
<span id="cb19-26"><a href="#cb19-26" tabindex="-1"></a></span>
<span id="cb19-27"><a href="#cb19-27" tabindex="-1"></a><span class="co"># Use before fitting</span></span>
<span id="cb19-28"><a href="#cb19-28" tabindex="-1"></a><span class="fu">validate_covariance_matrix</span>(my_cov_matrix, data)</span></code></pre></div>
<hr />
</div>
</div>
<div id="troubleshooting" class="section level2">
<h2>Troubleshooting</h2>
<div id="matrix-not-positive-semi-definite" class="section level3">
<h3>Matrix Not Positive Semi-Definite</h3>
<p><strong>Problem</strong>: Negative eigenvalues</p>
<p><strong>Solutions</strong>:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a><span class="co"># 1. Add small value to diagonal</span></span>
<span id="cb20-2"><a href="#cb20-2" tabindex="-1"></a>mat_fixed <span class="ot">&lt;-</span> mat <span class="sc">+</span> <span class="fu">diag</span>(<span class="fl">0.01</span>, <span class="fu">nrow</span>(mat))</span>
<span id="cb20-3"><a href="#cb20-3" tabindex="-1"></a></span>
<span id="cb20-4"><a href="#cb20-4" tabindex="-1"></a><span class="co"># 2. Nearest positive semi-definite matrix</span></span>
<span id="cb20-5"><a href="#cb20-5" tabindex="-1"></a><span class="fu">library</span>(Matrix)</span>
<span id="cb20-6"><a href="#cb20-6" tabindex="-1"></a>mat_fixed <span class="ot">&lt;-</span> <span class="fu">nearPD</span>(mat)<span class="sc">$</span>mat</span>
<span id="cb20-7"><a href="#cb20-7" tabindex="-1"></a></span>
<span id="cb20-8"><a href="#cb20-8" tabindex="-1"></a><span class="co"># 3. Check for errors in calculation</span></span></code></pre></div>
</div>
<div id="poor-convergence" class="section level3">
<h3>Poor Convergence</h3>
<p><strong>Problem</strong>: Model doesn’t converge</p>
<p><strong>Solutions</strong>: - Simplify structure (fewer
off-diagonals) - Scale matrix (diagonal = 1) - More iterations - Check
if structure is appropriate for data</p>
</div>
<div id="computational-speed" class="section level3">
<h3>Computational Speed</h3>
<p><strong>Problem</strong>: Large matrices slow</p>
<p><strong>Solutions</strong>:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a><span class="co"># Use sparse matrices</span></span>
<span id="cb21-2"><a href="#cb21-2" tabindex="-1"></a><span class="fu">library</span>(Matrix)</span>
<span id="cb21-3"><a href="#cb21-3" tabindex="-1"></a>sparse_cov <span class="ot">&lt;-</span> <span class="fu">as</span>(cov_matrix, <span class="st">&quot;dgCMatrix&quot;</span>)</span>
<span id="cb21-4"><a href="#cb21-4" tabindex="-1"></a></span>
<span id="cb21-5"><a href="#cb21-5" tabindex="-1"></a><span class="co"># Optimize mode</span></span>
<span id="cb21-6"><a href="#cb21-6" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">phybase_run</span>(..., <span class="at">optimise =</span> <span class="cn">TRUE</span>)</span>
<span id="cb21-7"><a href="#cb21-7" tabindex="-1"></a></span>
<span id="cb21-8"><a href="#cb21-8" tabindex="-1"></a><span class="co"># Reduce matrix size (subsample)</span></span></code></pre></div>
<hr />
</div>
</div>
<div id="see-also" class="section level2">
<h2>See Also</h2>
<ul>
<li><strong><a href="phylogenetic_models.html">Phylogenetic
Models</a></strong> - Special case of covariance structure</li>
<li><strong><a href="advanced_models.html">Advanced Models</a></strong>
- Random effects<br />
</li>
<li><strong><a href="getting_started.html">Getting Started</a></strong>
- Basic usage</li>
<li><strong><a href="model_diagnostics.html">Model
Diagnostics</a></strong> - Model checking</li>
</ul>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
